lib/std/fs.zig

```diff
@@ -594,8 +594,19 @@ pub const Dir = struct {
             const path_w = try os.windows.cStrToPrefixedFileW(sub_path);
             return self.openFileW(&path_w, flags);
         }
+
+        // Use the O_ locking flags if the os supports them
+        const lock_flag: u32 = lock_flag: {
+            if (!flags.lock) break :lock_flag 0;
+            if (flags.write) {
+                break :lock_flag if (@hasDecl(os, "O_EXLOCK")) os.O_EXLOCK else 0;
+            } else {
+                break :lock_flag if (@hasDecl(os, "O_SHLOCK")) os.O_SHLOCK else 0;
+            }
```

Comment:
Hmm... The only thing I can think of is perhaps having a program create a file, and then only keeping a read lock so that other processes can read from the file as it outputs to it. But at that point I don't know if the program would be using file locks at all.

Another scenario is that the process writes to a file, and then wants to keep holding a shared lock instead of an exclusive lock. In this case it would be nice if the file's lock type could be changed. However, `flock` doesn't support this usecase, and I'm not sure if windows does.

I guess the question is whether we should force people using the std file lock API to follow what the lock claims to be, or if they should be free to use the flags as they are implemented.

Reply:
Okay, looking at `flock`'s documentation a little closer:

- `LOCK_SH` vs `LOCK_EX` seems to deliberately leave out any mention of whether it is being written to or read from.
- I was wrong about `flock` not supporting changing locks. What it doesn't support is _atomic_ lock changes

On Windows the interface is explicitly about read, write, or delete access for other processes. I don't think it will stop from doing the scenario I suggested earlier.

With that in mind, I think it would be better to let the user select the type of lock explicitly. I think we could change the lock option to lock something like the following:

```zig
const Lock = enum {
    None
    Shared,
    Exclusive,
};

const OpenOptions = struct {
    lock: Lock = .None,
    // ... rest of OpenOptions excluded
};
```

- On Unix systems this will map directly to the `flock` options.
- On Windows this will map to differents sets of ShareAccess options:
   - `None => FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE`
   - `Shared => FILE_SHARE_READ`
   - `Shared => 0`

Reply:
Sounds good to me, let me know when it's ready for another look!